File: src/lib/simulator.ts
/**
 * XPS IRF Simulator - Main simulation engine
 * Ported from Python xps_twin package
 */

import {
  fermiDirac,
  skewGaussian,
  ellipticalGaussian2D,
  interp,
  convolve,
  gaussianKernel,
  linspace,
  meshgrid,
} from './physics';

export interface SimulatorParams {
  // X-ray Source parameters
  sigmaX: number;      // Spot size X (meV)
  sigmaY: number;      // Spot size Y (mm)
  alpha: number;       // Energy gradient (eV/mm)
  gammaX: number;      // Spot skewness X
  gammaY: number;      // Spot skewness Y

  // Detector parameters
  kappa: number;       // Smile curvature
  theta: number;       // Detector tilt (degrees)
  sigmaRes: number;    // Intrinsic resolution (eV)

  // Measurement parameters
  temp: number;        // Temperature (K)

  // Noise parameters (optional)
  poissonNoise?: number;
  gaussianNoise?: number;
}

export interface SimulationResult {
  energy: number[];           // Energy axis (eV)
  spectrum: number[];         // Observed spectrum (normalized)
  spectrumClean: number[];    // Spectrum without noise
  idealFD: number[];          // Ideal Fermi-Dirac
  irf: number[];              // Instrument Response Function
  image2D: number[][];        // 2D detector image
  spotProfile: number[][];    // 2D spot profile
  yAxis: number[];            // Y axis for 2D images
  // Resolution components (all in meV)
  sigmaSource: number;        // Source resolution
  sigmaDetector: number;      // Detector resolution
  sigmaCombined: number;      // Combined resolution (sqrt of sum of squares)
}

export interface Grid {
  eAxis: number[];
  yAxis: number[];
  E: number[][];
  Y: number[][];
  de: number;
}

/**
 * Create calculation grid
 */
function createGrid(
  eStart: number = -0.1,
  eEnd: number = 0.1,
  eSteps: number = 500,
  yStart: number = -10,
  yEnd: number = 10,
  ySteps: number = 200
): Grid {
  const eAxis = linspace(eStart, eEnd, eSteps);
  const yAxis = linspace(yStart, yEnd, ySteps);
  const { E, Y } = meshgrid(eAxis, yAxis);
  const de = eAxis[1] - eAxis[0];

  return { eAxis, yAxis, E, Y, de };
}

/**
 * Generate 2D emission from X-ray source
 */
function generate2DEmission(
  grid: Grid,
  trueSpectrum: number[],
  sigmaY: number,
  gammaY: number,
  alpha: number
): number[][] {
  const yDistribution = skewGaussian(grid.yAxis, sigmaY, gammaY);
  const rows = grid.yAxis.length;
  const cols = grid.eAxis.length;
  const img: number[][] = Array(rows).fill(null).map(() => Array(cols).fill(0));

  for (let i = 0; i < rows; i++) {
    const yVal = grid.yAxis[i];
    const shift = alpha * yVal;

    // Shift spectrum
    const shiftedEnergy = grid.eAxis.map(e => e - shift);
    const shiftedSpec = interp(shiftedEnergy, grid.eAxis, trueSpectrum, trueSpectrum[0], 0);

    for (let j = 0; j < cols; j++) {
      img[i][j] = shiftedSpec[j] * yDistribution[i];
    }
  }

  return img;
}

/**
 * Get 2D spot profile
 */
function get2DSpotProfile(
  grid: Grid,
  sigmaX: number,
  sigmaY: number,
  gammaX: number,
  gammaY: number
): number[][] {
  return ellipticalGaussian2D(grid.E, grid.Y, sigmaX, sigmaY, gammaX, gammaY, 0);
}

/**
 * Project 2D image to 1D spectrum through detector
 * Applies both source resolution (sigmaSource) and detector resolution (sigmaDetector)
 */
function projectTo1D(
  grid: Grid,
  image2D: number[][],
  kappa: number,
  theta: number,
  sigmaSource: number,
  sigmaDetector: number
): number[] {
  const thetaRad = (theta * Math.PI) / 180;
  const cosTheta = Math.cos(thetaRad);
  const sinTheta = Math.sin(thetaRad);

  const rows = grid.yAxis.length;
  const cols = grid.eAxis.length;
  const yMax = Math.max(...grid.yAxis.map(Math.abs));

  // Create distorted image via interpolation
  const distortedImg: number[][] = Array(rows).fill(null).map(() => Array(cols).fill(0));

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      const e = grid.E[i][j];
      const y = grid.Y[i][j];
      const yNorm = y / yMax;

      // Apply rotation and curvature
      const eSrc = e * cosTheta + y * sinTheta;
      const ySrc = -e * sinTheta + y * cosTheta;
      const eSrcCurved = eSrc - kappa * (yNorm * yNorm);

      // Bilinear interpolation
      const val = bilinearInterp(grid.yAxis, grid.eAxis, image2D, ySrc, eSrcCurved);
      distortedImg[i][j] = val;
    }
  }

  // Sum along Y axis
  let spec1D = new Array(cols).fill(0);
  for (let j = 0; j < cols; j++) {
    for (let i = 0; i < rows; i++) {
      spec1D[j] += distortedImg[i][j];
    }
  }

  // Apply source resolution (Gaussian convolution)
  if (sigmaSource > 0) {
    const kernelSource = gaussianKernel(sigmaSource, grid.de);
    spec1D = convolve(spec1D, kernelSource);
  }

  // Apply detector resolution (Gaussian convolution)
  if (sigmaDetector > 0) {
    const kernelDetector = gaussianKernel(sigmaDetector, grid.de);
    spec1D = convolve(spec1D, kernelDetector);
  }

  return spec1D;
}

/**
 * Bilinear interpolation for 2D array
 */
function bilinearInterp(
  yAxis: number[],
  xAxis: number[],
  data: number[][],
  y: number,
  x: number
): number {
  // Find indices
  const yMin = yAxis[0];
  const yMax = yAxis[yAxis.length - 1];
  const xMin = xAxis[0];
  const xMax = xAxis[xAxis.length - 1];

  if (y < yMin || y > yMax || x < xMin || x > xMax) {
    return 0;
  }

  const yStep = (yMax - yMin) / (yAxis.length - 1);
  const xStep = (xMax - xMin) / (xAxis.length - 1);

  const yi = (y - yMin) / yStep;
  const xi = (x - xMin) / xStep;

  const y0 = Math.floor(yi);
  const x0 = Math.floor(xi);
  const y1 = Math.min(y0 + 1, yAxis.length - 1);
  const x1 = Math.min(x0 + 1, xAxis.length - 1);

  const yt = yi - y0;
  const xt = xi - x0;

  const v00 = data[y0]?.[x0] ?? 0;
  const v01 = data[y0]?.[x1] ?? 0;
  const v10 = data[y1]?.[x0] ?? 0;
  const v11 = data[y1]?.[x1] ?? 0;

  return (
    v00 * (1 - xt) * (1 - yt) +
    v01 * xt * (1 - yt) +
    v10 * (1 - xt) * yt +
    v11 * xt * yt
  );
}

/**
 * Add noise to spectrum
 */
function addNoise(
  spectrum: number[],
  poissonLevel: number,
  gaussianLevel: number
): number[] {
  return spectrum.map(val => {
    let noisy = val;

    // Poisson noise (approximated as Gaussian with sqrt(N) std)
    if (poissonLevel > 1e-5) {
      const scaleFactor = 1000.0 / poissonLevel;
      const lambda = val * scaleFactor;
      // Box-Muller for Gaussian approximation of Poisson
      const u1 = Math.random();
      const u2 = Math.random();
      const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
      noisy = (lambda + z * Math.sqrt(Math.max(lambda, 0))) / scaleFactor;
    }

    // Gaussian noise
    if (gaussianLevel > 0) {
      const u1 = Math.random();
      const u2 = Math.random();
      const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
      noisy += z * (gaussianLevel / 100);
    }

    return Math.max(0, noisy);
  });
}

/**
 * Compute numerical gradient (for IRF extraction)
 */
function gradient(data: number[], dx: number): number[] {
  const n = data.length;
  const result = new Array(n).fill(0);

  // Central difference for interior points
  for (let i = 1; i < n - 1; i++) {
    result[i] = (data[i + 1] - data[i - 1]) / (2 * dx);
  }

  // Forward/backward difference for endpoints
  result[0] = (data[1] - data[0]) / dx;
  result[n - 1] = (data[n - 1] - data[n - 2]) / dx;

  return result;
}

/**
 * Main simulation function
 */
export function simulate(params: SimulatorParams): SimulationResult {
  // Create grid with extended range for proper edge handling
  // Internal calculation uses wider range (-0.15 to 0.15) to avoid edge artifacts
  const gridExtended = createGrid(-0.15, 0.15, 750, -10, 10, 200);
  // Display grid is the standard range
  const grid = createGrid(-0.1, 0.1, 500, -10, 10, 200);

  // Convert units: sigmaX (source resolution) and sigmaRes (detector resolution) are in meV, convert to eV
  const sigmaSourceEV = params.sigmaX / 1000;
  const sigmaDetectorEV = params.sigmaRes / 1000;

  // Calculate combined resolution: sqrt(sigma_source^2 + sigma_detector^2)
  const sigmaCombinedMeV = Math.sqrt(params.sigmaX ** 2 + params.sigmaRes ** 2);

  // 1. Compute ideal Fermi-Dirac distribution on extended grid
  const idealFDExtended = fermiDirac(gridExtended.eAxis, params.temp);

  // 2. Generate 2D emission from X-ray source on extended grid
  const image2DExtended = generate2DEmission(
    gridExtended,
    idealFDExtended,
    params.sigmaY,
    params.gammaY,
    params.alpha
  );

  // 3. Get 2D spot profile (for display, use standard grid)
  const spotProfile = get2DSpotProfile(
    grid,
    sigmaSourceEV,
    params.sigmaY,
    params.gammaX,
    params.gammaY
  );

  // 4. Project to 1D through detector on extended grid
  //    Apply both source and detector resolution convolutions
  const spectrumRawExtended = projectTo1D(
    gridExtended,
    image2DExtended,
    params.kappa,
    params.theta,
    sigmaSourceEV,
    sigmaDetectorEV
  );

  // 5. Trim to display range and interpolate
  const spectrumRaw = interp(grid.eAxis, gridExtended.eAxis, spectrumRawExtended);

  // 6. Normalize
  const maxVal = Math.max(...spectrumRaw);
  const spectrumClean = spectrumRaw.map(v => v / (maxVal + 1e-12));

  // 7. Add noise
  const spectrum = addNoise(
    spectrumClean,
    params.poissonNoise ?? 0,
    params.gaussianNoise ?? 0
  );

  // 8. Compute ideal FD for display range
  const idealFD = fermiDirac(grid.eAxis, params.temp);

  // 9. Extract IRF (simulate at very low temperature and differentiate)
  const stepFD = fermiDirac(gridExtended.eAxis, 0.01);
  const stepImage = generate2DEmission(gridExtended, stepFD, params.sigmaY, params.gammaY, params.alpha);
  const stepSpectrumExtended = projectTo1D(
    gridExtended,
    stepImage,
    params.kappa,
    params.theta,
    sigmaSourceEV,
    sigmaDetectorEV
  );
  const stepSpectrum = interp(grid.eAxis, gridExtended.eAxis, stepSpectrumExtended);

  const irfRaw = gradient(stepSpectrum, grid.de);
  const irfMax = Math.max(...irfRaw.map(Math.abs));
  const irf = irfRaw.map(v => -v / (irfMax + 1e-12));  // Negative for BE direction

  // 10. Generate 2D image for display (use display grid)
  const image2D = generate2DEmission(
    grid,
    idealFD,
    params.sigmaY,
    params.gammaY,
    params.alpha
  );

  return {
    energy: grid.eAxis.map(e => e * 1000),  // Convert to meV
    spectrum,
    spectrumClean,
    idealFD,
    irf,
    image2D,
    spotProfile,
    yAxis: grid.yAxis,
    // Resolution components (in meV)
    sigmaSource: params.sigmaX,
    sigmaDetector: params.sigmaRes,
    sigmaCombined: sigmaCombinedMeV,
  };
}

/**
 * Default parameters
 */
export const defaultParams: SimulatorParams = {
  sigmaX: 0.5,
  sigmaY: 0.5,
  alpha: 0.002,
  gammaX: 0.0,
  gammaY: 0.0,
  kappa: 0.01,
  theta: 0.08,
  sigmaRes: 1.5,
  temp: 5.0,
  poissonNoise: 2.0,
  gaussianNoise: 1.0,
};

File: fastapi-htmx/app/simulator.py
"""
XPS IRF Simulator - Main simulation engine.
Generates simulated XPS spectra with instrumental response function effects.
"""

import numpy as np
from dataclasses import dataclass
from typing import Tuple
from scipy.ndimage import map_coordinates

from .physics import (
    fermi_dirac,
    skew_gaussian,
    convolve_gaussian,
    elliptical_gaussian_2d,
    numerical_gradient,
)


@dataclass
class SimulationParams:
    """Simulation parameters."""
    # X-ray source
    sigma_source: float = 20.0   # Source resolution (meV)
    sigma_spot: float = 2.0      # Spot width (mm)
    gamma_energy: float = 0.0    # Energy skewness
    gamma_spatial: float = 0.0   # Spatial skewness
    alpha: float = 0.0           # Energy gradient (eV/mm)

    # Detector
    kappa: float = 0.0           # Smile curvature
    theta: float = 0.0           # Detector tilt (degrees)
    sigma_detector: float = 10.0 # Detector resolution (meV)

    # Measurement
    temperature: float = 300.0   # Temperature (K)
    poisson_noise: float = 4.0   # Poisson noise level (log scale)
    gaussian_noise: float = 1.0  # Gaussian noise (%)


@dataclass
class SimulationResult:
    """Simulation result container."""
    energy: np.ndarray           # Energy axis (meV)
    spectrum: np.ndarray         # Observed spectrum (with noise)
    spectrum_clean: np.ndarray   # Clean spectrum (no noise)
    ideal_fd: np.ndarray         # Ideal Fermi-Dirac
    irf: np.ndarray              # Instrumental response function
    image_2d: np.ndarray         # 2D detector image
    spot_profile: np.ndarray     # 2D spot profile
    y_axis: np.ndarray           # Spatial axis (mm)
    sigma_combined: float        # Combined resolution (meV)


# Grid configuration
ENERGY_POINTS = 500
SPATIAL_POINTS = 200
ENERGY_RANGE = (-0.1, 0.1)      # eV (-100 to 100 meV)
SPATIAL_RANGE = (-10.0, 10.0)   # mm
EXTENDED_ENERGY_RANGE = (-0.15, 0.15)  # Extended range for convolution


def create_grids() -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """Create energy and spatial grids."""
    # Extended energy grid for internal calculations
    energy_ext = np.linspace(EXTENDED_ENERGY_RANGE[0], EXTENDED_ENERGY_RANGE[1],
                             int(ENERGY_POINTS * 1.5))
    # Display energy grid
    energy = np.linspace(ENERGY_RANGE[0], ENERGY_RANGE[1], ENERGY_POINTS)
    # Spatial grid
    y_axis = np.linspace(SPATIAL_RANGE[0], SPATIAL_RANGE[1], SPATIAL_POINTS)

    return energy, energy_ext, y_axis, energy


def simulate(params: SimulationParams) -> SimulationResult:
    """
    Run XPS simulation with given parameters.

    Args:
        params: Simulation parameters

    Returns:
        SimulationResult with all computed arrays
    """
    # Create grids
    energy, energy_ext, y_axis, _ = create_grids()
    de = energy[1] - energy[0]  # Energy step
    dy = y_axis[1] - y_axis[0]  # Spatial step

    # Convert parameters to eV
    sigma_src_ev = params.sigma_source / 1000.0
    sigma_det_ev = params.sigma_detector / 1000.0
    sigma_combined = np.sqrt(sigma_src_ev**2 + sigma_det_ev**2) * 1000  # back to meV

    # Step 1: Ideal Fermi-Dirac distribution
    ideal_fd = fermi_dirac(energy, params.temperature)

    # Step 2: Generate 2D emission (energy × spatial)
    # Create meshgrid
    E, Y = np.meshgrid(energy_ext, y_axis, indexing='ij')

    # Apply energy gradient: shift spectrum by alpha * y
    image_2d = np.zeros((len(energy_ext), len(y_axis)))

    for j, y in enumerate(y_axis):
        # Energy shift due to gradient
        e_shifted = energy_ext - params.alpha * y
        # Fermi-Dirac at shifted energy
        fd_shifted = fermi_dirac(e_shifted, params.temperature)
        # Spatial distribution (skew Gaussian)
        spatial_weight = skew_gaussian(np.array([y]), params.sigma_spot, params.gamma_spatial)[0]
        image_2d[:, j] = fd_shifted * spatial_weight

    # Step 3: Generate spot profile for visualization
    E_spot, Y_spot = np.meshgrid(energy, y_axis, indexing='ij')
    spot_profile = elliptical_gaussian_2d(
        E_spot * 1000,  # Convert to meV for visualization
        Y_spot,
        sigma_src_ev * 1000,
        params.sigma_spot,
        params.gamma_energy,
        params.gamma_spatial
    )

    # Step 4: Apply detector distortions (smile + tilt)
    theta_rad = np.deg2rad(params.theta)
    cos_t, sin_t = np.cos(theta_rad), np.sin(theta_rad)

    # Distorted image coordinates
    distorted_image = np.zeros_like(image_2d)

    for i, e in enumerate(energy_ext):
        for j, y in enumerate(y_axis):
            # Rotation (tilt)
            e_rot = e * cos_t + (y / 100) * sin_t  # Scale y to similar magnitude
            y_rot = -e * sin_t + (y / 100) * cos_t

            # Smile curvature: parabolic distortion
            y_norm = y / SPATIAL_RANGE[1]  # Normalize to [-1, 1]
            e_curved = e_rot - params.kappa * y_norm**2 * 0.01  # Scale curvature effect

            # Find source coordinates (inverse mapping)
            # Simple nearest-neighbor for now
            e_idx = int((e_curved - EXTENDED_ENERGY_RANGE[0]) / de)
            if 0 <= e_idx < len(energy_ext):
                distorted_image[i, j] = image_2d[e_idx, j]

    # Step 5: Project to 1D spectrum (sum over spatial axis)
    spectrum_1d_ext = distorted_image.sum(axis=1)

    # Step 6: Apply resolution broadening (convolution)
    # Source resolution
    spectrum_1d_ext = convolve_gaussian(spectrum_1d_ext, sigma_src_ev, de)
    # Detector resolution
    spectrum_1d_ext = convolve_gaussian(spectrum_1d_ext, sigma_det_ev, de)

    # Interpolate to display grid
    spectrum_clean = np.interp(energy, energy_ext, spectrum_1d_ext)

    # Normalize
    if spectrum_clean.max() > 0:
        spectrum_clean = spectrum_clean / spectrum_clean.max()

    # Step 7: Add noise
    spectrum = add_noise(spectrum_clean.copy(), params.poisson_noise, params.gaussian_noise)

    # Step 8: Extract IRF (derivative of step response at T≈0)
    # Simulate with very low temperature
    step_params = SimulationParams(
        sigma_source=params.sigma_source,
        sigma_spot=params.sigma_spot,
        gamma_energy=params.gamma_energy,
        gamma_spatial=params.gamma_spatial,
        alpha=params.alpha,
        kappa=params.kappa,
        theta=params.theta,
        sigma_detector=params.sigma_detector,
        temperature=0.01,  # Near zero
        poisson_noise=10,  # Low noise
        gaussian_noise=0,
    )

    # For IRF, we need the step response (simplified calculation)
    step_fd = fermi_dirac(energy, 0.01)
    step_spectrum = convolve_gaussian(step_fd, sigma_src_ev, de)
    step_spectrum = convolve_gaussian(step_spectrum, sigma_det_ev, de)

    # IRF = -d(step)/dE
    irf = -numerical_gradient(step_spectrum, de)
    # Normalize IRF
    if irf.max() > 0:
        irf = irf / irf.max()

    # Prepare 2D detector image for display (interpolate to display grid)
    image_display = np.zeros((len(energy), len(y_axis)))
    for j in range(len(y_axis)):
        image_display[:, j] = np.interp(energy, energy_ext, distorted_image[:, j])

    return SimulationResult(
        energy=energy * 1000,  # Convert to meV for display
        spectrum=spectrum,
        spectrum_clean=spectrum_clean,
        ideal_fd=ideal_fd,
        irf=irf,
        image_2d=image_display,
        spot_profile=spot_profile,
        y_axis=y_axis,
        sigma_combined=sigma_combined,
    )


def add_noise(spectrum: np.ndarray, poisson_level: float, gaussian_level: float) -> np.ndarray:
    """
    Add Poisson and Gaussian noise to spectrum.

    Args:
        spectrum: Clean spectrum
        poisson_level: Poisson noise level (log scale, higher = less noise)
        gaussian_level: Gaussian noise level (%)

    Returns:
        Noisy spectrum
    """
    # Poisson noise (approximated with Gaussian for high counts)
    if poisson_level < 10:
        scale = 10 ** poisson_level
        # Poisson-like noise using normal approximation
        counts = spectrum * scale
        noise = np.random.normal(0, np.sqrt(np.maximum(counts, 1)))
        spectrum = (counts + noise) / scale

    # Gaussian noise
    if gaussian_level > 0:
        noise = np.random.normal(0, gaussian_level / 100, len(spectrum))
        spectrum = spectrum + noise

    return np.clip(spectrum, 0, None)
